### ISSUE TITLE
nsqd: invalid diskqueue message sizes

### ISSUE THREAD

**Issue by twmb** [bitly/nsq#598]
If a backing message is corrupt and has a message size of 0, diskQueue.readOne will return empty messages forever.

https://github.com/bitly/nsq/blob/a44b475b3f55e9f2e8247223b0b89f1029a30668/nsqd/diskqueue.go#L257

This should enforce msgSize of at least 1 byte.

I do not know how a message was written with 0 msgSize as this is the first time this has happened.

**Comment by twmb** [comment]
Actually this is incorrect. A msgSize of 0 will advance the readPos 4 bytes on binary.Read. However, this code should enforce that the message size is at least 26 (as message.go's decodeMessage does).  If <26, this is an invalid message, and there is no reasonable guarantee about where any message in the current disk segment is, and the segment itself should be deleted. Invalid message reads may and will eventually get to an int32 that is <0, and cause nsqd to panic.

Currently, all uses of ReadChan currently return []byte encoded nsq.Message's. It might make sense to decode messages in readOne.

It also might be good to have a checksum of message encoded when written to disk, so that if msgSize is >26, we can validate the message contents to ensure that msgSize was not written incorrectly.

**Comment by mreiferson** [comment]
I think a minimal and useful improvement is to just check the size as you've suggested inline in `readOne`:

```
if msgSize < 26 || msgSize > ...opts.maxMsgSize {
    ...
}
```

@twmb want to submit a patch?

> It also might be good to have a checksum of message encoded when written to disk, so that if msgSize is >26, we can validate the message contents to ensure that msgSize was not written incorrectly.

Future persistence related improvements will include checksums, I regret not adding them :(

**Comment by mreiferson** [comment]
fixed in #600
